Add README (#90)
wavefunction91/IntegratorXX: README.md

# IntegratorXX

## Synopsis

> IntegratorXX is a modern C++ library for the generation of atomic and molecular
> grids for quantum chemistry calculations. It provides a uniform interface for
> the generation of primitive, radial and solid angle quadratures, as well as
> there combination into spherical grids.

Reviewer: ```suggestion

IntegratorXX is a modern C++ library for the generation of atomic and molecular
grids for quantum chemistry calculations. The most important use of these
quadratures is the calculation of the exchange-correlation energy in density
functional theory

$$ E_{\text{xc}}=\int f_{\text{xc}}(n,\nabla n,\nabla^{2}n,\tau){\rm d}^{3}r $$

using the multi-center approach pioneered by Becke in [J. Chem. Phys. 88,
2547–2553 (1988)](https://doi.org/10.1063/1.454033); however, the
quadratures are also useful in many other contexts.

IntegratorXX provides a uniform interface for the generation of 
primitive, radial and solid angle quadratures, as well as
their combination into spherical grids.
```

Author: I don't think it's a good idea to pin ourselves to XC-related topics - as you've said yourself, grids like these are required to evaluate _**everything**_ in other basis expansions. You want to list some examples, that's fine, but going into things like the Becke scheme for the purposes of a README is a bit pedantic IMHO.

Reviewer: Well yes, but given that XC is going to be 99% of our usage, I think it would be silly not to mention it here :)

## Design Goals 

> * Provide a stable and reproducible implementation of the various atomic and
> molecular grids commonly encountered in quantum chemistry calculations

Reviewer: ```suggestion
* Provide stable, reusable, and reproducible implementations of the various atomic and
molecular grids commonly encountered in quantum chemistry calculations
```

* Develop a modern, modular, extensible C++ API to allow for the implementation
and validation new atomic and molecular quadrature schemes.

> 

Reviewer: ```suggestion
* Provide complete C and Python interfaces to allow reusing the implementation in projects written in these languages
```

| Gauss-Chebyshev (First Kind)    | Primitive        | (-1,1)  | `GaussChebyshev1`   |
| Gauss-Chebyshev (Second Kind)   | Primitive        | [-1,1]  | `GaussChebyshev2`   |
| Gauss-Chebyshev (Third Kind)    | Primitive        | (0,1)   | `GaussChebyshev3`   |
| Gauss-Legendre                  | Primitive        | [-1,1]  | `GaussLegendre`     |
| Gauss-Lobatto                   | Primitive        | [-1,1]  | `GaussLobatto`      |
| Trapezoid Rule                  | Primitive        | [0,1]   | `UniformTrapezoid`  |
| Becke                           | Radial           | (0,inf) | `Becke`             |
| Murray-Handy-Laming (MHL)       | Radial           | (0,inf) | `MurrayHandyLaming` |
| Mura-Knowles (MK)               | Radial           | (0,inf) | `MuraKnowles`       |
| Treutler-Ahlrichs (TA, M3 + M4) | Radial           | (0,inf) | `TreutlerAhlrichs`  |
| Ahrens-Beylkin                  | Angular          | $S^2$   | `AhrensBeylkin`     |
| Delley                          | Angular          | $S^2$   | `Delley`            |
| Lebedev-Laikov                  | Angular          | $S^2$   | `LebedevLaikov`     |

> | Womersley                       | Angular          | $S^2$   | `Womersley`         |

Reviewer: I would split this table in three: primitive, radial, and angular quadratures, which could be discussed in subsections.

Author: Agreed. Fixed

For the generation of spherical quadratures, IntegratorXX additionally supports the following radial pruning schemes:
| Name      | Description                          | C++ Specifier             |
|-----------|--------------------------------------|---------------------------|

> | Unpruned  | No not perform pruning               | `PruningScheme::Unpruned` |

Reviewer: ```suggestion
| Unpruned  | Do not perform pruning               | `PruningScheme::Unpruned` |
```

we provide a set of look-up functions which can safely produce compatible grid orders:

```
using angular_type = LebedevLaikov<double>; // FP64 LL grid, similar for other implementations
using traits = IntegratorXX::quadrature_traits<angular_type>;

auto npts  = traits::npts_by_algebraic_order(order); // Return the grid order associated with a particular algebraic order
auto order = traits::algebraic_order_by_npts(npts);  // Return the algebratic order associated with a grid order
auto next_order = traits::next_algebraic_order(order); // Return the next largest (inclusive) algebratic order compatible with `angular_type` 
```

## Example Usage

Many example usages for 1-d quadratures (i.e. primitive and radial) can be found in `test/1d_quadratures.cxx` and `test/spherical_generator.cxx`. Below is a simple invocation example for the generation of an atomic sphere via the runtime generator:
```
using namespace IntegratorXX;               // Import namespace
auto rad_spec = RadialQuad::MuraKnowles;    // MK Radial scheme
auto ang_spec = AngularQuad::AhrensBeylkin; // AH Angular scheme
size_t nrad   = 99;
size_t nang   = 372;

> double rscal  = 2.0;

Reviewer: My issue with this is that this requires hardcoding the allowed options. I would like to have alternative functionality, where the library takes in a string specification for the rule and returns the radial or angular specification. That way the downstream interfaces don't have to be changed when new rules are added in IntegratorXX. If this is already possible, please provide an example

Author: Based on where you placed this comment, I'm unclear what you mean here - do you mean the specification of the grid orders / scaling factors or the usage of `enums` to specify the the radial / angular rules? 

If you want a string-driven "getter" of the radial/angular specifications, that's fine, but the internal implementations  _**must**_ be discrete, as we have to generate the grids using compile time logic. For this purpose, `enum` == `int` (you'll also notice that I manually specify the integer encodings rather than letting them be auto generate - this is looking forward to making the C/C++ interfaces compatible).

I'd be fine with something like the following:
```
RadialQuad radial_scheme_from_string(std::string name);
AngularQuad angular_scheme_from_string(std::string name);
...
auto rad_spec = radial_scheme_from_string("MuraKnowles");
auto ang_spec = angular_scheme_from_string("AhrensBeylkin");
```
Would that resolve the issue you raise?

Author: Added as of https://github.com/wavefunction91/IntegratorXX/pull/90/commits/90a855fb3d900c3bd7850a22f39e9b513b09b273

Reviewer: Yes. I was just aiming at it being very hard to extend hardcoded implementations. Having the parser logic in the library makes downstream codes future proof, since you can even allow user-side specification of the input string, kind of how functionals are (or should be!) exlusively handled by libxc...

// Generate Grid Specification
UnprunedSphericalGridSpecification unp{
  rad_spec, nrad, rscal, ang_spec, nang
};

> auto pruning_spec = create_pruned_spec(PruningScheme::Robust, unp 

Reviewer: broken code

Author: Fixed

source for appropriate references.

### Primitive Quadratures

Primitive quadratures are those generated on a finite bound (e.g. Gauss
quadrature rules). The general software design pattern of IntegratorXX is to
build up higher-order quadrature rules (e.g. radial transformation, etc) from
these primitive quadratures.

| Quadrature Name                 | Domain    | C++ Class           |
|---------------------------------|-----------|---------------------|
| Gauss-Chebyshev (First Kind)    | $(-1,1)$  | `GaussChebyshev1`   |
| Gauss-Chebyshev (Second Kind)   | $[-1,1]$  | `GaussChebyshev2`   |
| Gauss-Chebyshev (Third Kind)    | $(0,1)$   | `GaussChebyshev3`   |
| Gauss-Legendre                  | $[-1,1]$  | `GaussLegendre`     |
| Gauss-Lobatto                   | $[-1,1]$  | `GaussLobatto`      |
| Trapezoid Rule                  | $[0,1]$   | `UniformTrapezoid`  |

### Radial Quadratures

Radial quadratures are convolutions of primitive quadrature rules with a radial
transformation scheme (mapping the natural domain of the primitive quadrature
to positive semi-indefinite). The jacobian of the transformation *is* included

> in the radial weights while the radial component of the spherical jacobian

Reviewer: should one say "volume element" instead of jacobian here..?

Author: That's fair - I typically use them interchangeably as the meaning is to be understood by context, but I'd agree that volume element is the strictly correct term here. Updated

($r^2$) is *not*.

| Quadrature Name                 | Domain       | C++ Class           |
|---------------------------------|--------------|---------------------|
| Becke                           | $(0,\infty)$ | `Becke`             |
| Murray-Handy-Laming (MHL)       | $(0,\infty)$ | `MurrayHandyLaming` |
| Mura-Knowles (MK)               | $(0,\infty)$ | `MuraKnowles`       |
| Treutler-Ahlrichs (TA, M3 + M4) | $(0,\infty)$ | `TreutlerAhlrichs`  |

### Angular Quadratures

> Angular quadratures integrate over $S^2$ (solid angle). It is typically the
> case that these are manually constructed to integrate spherical harmonics up to
> a specific order, and are thus only compatible witch *specific* grid orders (see
> note below).

Reviewer: ```suggestion
Angular quadratures integrate over $S^2$ (solid angle). These have typically been
manually constructed to integrate spherical harmonics up to a specific order, and
are thus only compatible witch *specific* grid orders (see note below).
```
